File: schemas/content-moderation/drizzle/moderation_action_log.ts
// moderation_action_log: Immutable audit trail of moderation events.
// See README.md for full design rationale.

import {
  pgTable,
  uuid,
  text,
  jsonb,
  timestamp,
  index,
} from "drizzle-orm/pg-core";
import { users } from "../../auth-rbac/drizzle/users";

export const moderationActionLog = pgTable(
  "moderation_action_log",
  {
    id: uuid("id").primaryKey().defaultRandom(),

    actorId: uuid("actor_id")
      .notNull()
      .references(() => users.id, { onDelete: "restrict" }),

    actionType: text("action_type").notNull(), // What happened. Not an enum — new action types should not require schema migration.
    targetType: text("target_type").notNull(), // What entity the action was on (e.g., "queue_item", "report", "user", "moderation_rule", "policy").
    targetId: text("target_id").notNull(), // ID of the target entity.

    details: jsonb("details"), // Event-specific context (e.g., action_taken: {"action_type": "ban", "reason": "...", "duration": "24h"}).
    ipAddress: text("ip_address"), // Client IP address for security audit.

    // Immutable. Actions are append-only — no updated_at.
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("idx_moderation_action_log_actor_id").on(table.actorId),
    index("idx_moderation_action_log_action_type").on(table.actionType),
    index("idx_moderation_action_log_target").on(table.targetType, table.targetId),
    index("idx_moderation_action_log_created_at").on(table.createdAt),
  ]
);

File: schemas/content-moderation/spacetimedb/moderation_action_log.rs
// moderation_action_log: Immutable audit trail of moderation events.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = moderation_action_log, public)]
pub struct ModerationActionLog {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub actor_id: String, // FK → users.id (restrict delete)

    #[index(btree)]
    pub action_type: String, // What happened (e.g., "report_created", "action_taken", "appeal_decided", "rule_updated").
                             // Intentionally not an enum — new action types should not require schema migration.

    #[index(btree)]
    pub target_type: String, // What entity the action was on (e.g., "queue_item", "report", "user", "moderation_rule", "policy").
                             // Intentionally not an enum — same rationale as action_type.

    #[index(btree)]
    pub target_id: String, // ID of the target entity. Not a FK — target may be any entity type.

    pub details: Option<String>, // JSON. Event-specific context (e.g., action_taken: {"action_type": "ban", "reason": "...", "duration": "24h"}).

    pub ip_address: Option<String>, // Client IP address for security audit.

    #[index(btree)]
    pub created_at: Timestamp, // Immutable. Actions are append-only — no updated_at.
}
